src/providers/BackgroundReplacementProvider/index.tsx

```diff
@@ -0,0 +1,170 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import{
+  BackgroundReplacementOptions,
+  BackgroundReplacementVideoFrameProcessor,
+  BackgroundFilterSpec,
+  ConsoleLogger,
+  DefaultVideoTransformDevice,
+  Device,
+  LogLevel,
+  NoOpVideoFrameProcessor,
+  VideoFrameProcessor,
+  BackgroundReplacementProcessor,
+} from 'amazon-chime-sdk-js';
+import React, {
+  createContext,
+  FC,
+  useContext,
+  useEffect,
+  useState,
+} from 'react';
+import useMemoCompare from '../../utils/use-memo-compare';
+import { isPrevNextUndefined, isPrevNextEmpty } from '../../utils/device-utils';
+import { BaseSdkProps } from '../../components/sdk/Base';
+
+interface Props extends BaseSdkProps {
+  /** The spec defines the assets that will be used for adding background replacement to a frame. For more information, refer to
+   * [Amazon Chime SDK for JavaScript Background Filter Guide](https://github.com/aws/amazon-chime-sdk-js/blob/main/guides/15_Background_Filter_Video_Processor.md#adding-a-background-filter-to-your-application). */
+  spec?: BackgroundFilterSpec;
+  /* A set of options that can be supplied when creating a background replacement video frame processor such as the background replacement image blob. For more information, refer to
+   * [Amazon Chime SDK for JavaScript Background Filter Guide](https://github.com/aws/amazon-chime-sdk-js/blob/main/guides/15_Background_Filter_Video_Processor.md#adding-a-background-filter-to-your-application). */
+  options?: BackgroundReplacementOptions;
+}
+
+interface BackgroundReplacementProviderState {
+  createBackgroundReplacementDevice: (
+    device: Device
+  ) => Promise<DefaultVideoTransformDevice>;
+  isBackgroundReplacementSupported: boolean | undefined;
+}
+
+const BackgroundReplacementProviderContext =
+  createContext<BackgroundReplacementProviderState | undefined>(undefined);
+
+const BackgroundReplacementProvider: FC<Props> = ( { spec, options, children }) => {
+  const [isBackgroundReplacementSupported, setIsBackgroundReplacementSupported] = useState<boolean | undefined>(undefined);
+  const [processor, setProcessor] = useState<VideoFrameProcessor | undefined>(undefined);
+  
+  const replacementSpec = useMemoCompare(
+    spec,
+    (
+      prev: BackgroundFilterSpec | undefined,
+      next: BackgroundFilterSpec | undefined
+    ): boolean => {
+      if (Object.is(prev, next)) {
+        return true;
+      }
+      return false;
+    }
+  );
+
+  const replacementOptions = useMemoCompare(
+    options,
+    (
+      prev: BackgroundReplacementOptions | undefined,
+      next: BackgroundReplacementOptions | undefined
+    ): boolean => {
+      if (isPrevNextUndefined(prev, next) || isPrevNextEmpty(prev, next) || (prev?.imageBlob?.size === next?.imageBlob?.size && prev?.filterCPUUtilization === next?.filterCPUUtilization)) {
```

Comment:
I just noticed the following after breaking down these helper methods.

Actually, this check

```
prev?.imageBlob?.size === next?.imageBlob?.size && prev?.filterCPUUtilization === next?.filterCPUUtilization
```

covers the following:

```
isPrevNextUndefined(prev, next) || isPrevNextEmpty(prev, next)
```

For example, suppose that `prev` is `undefined` or an empty object `{}`. `prev?.imageBlob?.size ` will be `undefined`. Please correct me if I am missing something.

Reply:
I talked with Hekmat that we'll refactor after release because `isPrevNextUndefined` and `isPrevNextEmpty` work as expected. They are also explicit.